File: ios/ExpoKakaoAdModule.swift
import ExpoModulesCore
import UIKit
import KakaoAdSDK
import AdSupport
import AppTrackingTransparency

struct KakaoAdDetailProduct: Record {
  @Field
  var id: String = ""
  
  @Field
  var name: String = ""
  
  @Field
  var quantity: Int = 0
  
  @Field
  var price: Double = 0.0
}

public class ExpoKakaoAdModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoKakaoAd")

        /*
         * 광고계정별로 고유한 식별값. trackId 를 설정하지 않을 경우 이벤트 수집은 이루어지지 않습니다.
         */
        OnCreate {
            if let trackId = Bundle.main.object(forInfoDictionaryKey: "KAKAO_AD_TRACK_ID") as? String {
                KakaoAdTracker.trackId = trackId
            }
        }

        /*
         * 이벤트 트래커 활성화
         */
        AsyncFunction("activate") { (promise: Promise) in
            if #available(iOS 14, *) {
                ATTrackingManager.requestTrackingAuthorization { status in
                    switch status {
                    case .authorized:
                        // 추적 권한 요청 팝업이 뜨고, 사용자가 허용함.
                        print("Authorized")
                        // 광고식별자 확인 가능
                        print(ASIdentifierManager.shared().advertisingIdentifier)
                    case .denied:
                        // 사용자가 거부함.
                        print("Denied")
                    case .notDetermined:
                        // 팝업이 뜨지 않음.
                        print("Not Determined")
                    case .restricted:
                        print("Restricted")
                    @unknown default:
                        print("Unknown")
                    }
                    // 권한 요청 완료 후 트래커 활성화
                    KakaoAdTracker.activate()
                    promise.resolve()
                }
            } else {
                // iOS 14 미만에서는 바로 활성화
                KakaoAdTracker.activate()
                promise.resolve()
            }
        }

        /*
         * 회원가입 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendCompleteRegistrationEvent") { (tag: String?) in
            KakaoAdTracker.sendCompleteRegisterEvent(tag: tag)
        }

        /*
         * 로그인 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendLoginEvent") { (tag: String?) in
            KakaoAdTracker.sendLoginEvent(tag: tag)
        }

        /*
         * Search Event
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         * @param searchString: 검색어
         */
        Function("sendSearchEvent") { (tag: String?, searchString: String?) in
            KakaoAdTracker.sendSearchEvent(tag: tag, searchString: searchString)
        }

        /*
         * 콘텐츠/상품 조회 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         * @param contentId: 콘텐츠/상품 코드
         */
        Function("sendViewContentEvent") { (tag: String?, contentId: String?) in
            KakaoAdTracker.sendViewContentEvent(tag: tag, contentId: contentId)
        }

        /*
         * 관심상품 추가 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         * @param contentId: 콘텐츠/상품 코드
         */
        Function("sendAddToWishListEvent") { (tag: String?, contentId: String?) in
            KakaoAdTracker.sendAddToWishListEvent(tag: tag, contentId: contentId)
        }

        /*
         * 장바구니 추가 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         * @param contentId: 콘텐츠/상품 코드
         */
        Function("sendAddToCartEvent") { (tag: String?, contentId: String?) in
            KakaoAdTracker.sendAddToCartEvent(tag: tag, contentId: contentId)
        }

        /*
         * 장바구니 보기 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendViewCartEvent") { (tag: String?) in
            KakaoAdTracker.sendViewCartEvent(tag: tag)
        }
        
        /*
         * 구매 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         * @param totalQuantity: 주문에 포함된 상품수
         * @param totalPrice: 주문 전체 금액
         * @param currency: 결제 금액 통화 (ISO4217 포맷)
         * @param products: KakaoAdDetailProduct 클래스 객체의 Array 타입을 입력 받습니다. 개별 주문 상품 상세 정보의 리스트 입니다.
         */
        Function("sendPurchaseEvent") { (tag: String?, totalQuantity: Int, totalPrice: Double, currency: String?, products: [KakaoAdDetailProduct]) in
            let kakaoProducts = products.map { product in
                KakaoAdSDK.KakaoAdDetailProduct(
                    id: product.id,
                    name: product.name,
                    quantity: product.quantity,
                    price: product.price
                )
            }
            
            KakaoAdTracker.sendPurchaseEvent(
                tag: tag,
                totalQuantity: totalQuantity,
                totalPrice: totalPrice,
                currency: currency,
                products: kakaoProducts
            )
        }
       
        /*
         * 잠재고객 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendParticipationEvent") { (tag: String?) in
            KakaoAdTracker.sendParticipationEvent(tag: tag)
        }

        /*
         * 사전준비 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendPreparationEvent") { (tag: String?) in
            KakaoAdTracker.sendPreparationEvent(tag: tag)
        }

        /*
         * 서비스신청 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendSignUpEvent") { (tag: String?) in
            KakaoAdTracker.sendSignUpEvent(tag: tag)
        }
        
        /*
         * 튜토리얼 이벤트
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendTutorialEvent") { (tag: String?) in
            KakaoAdTracker.sendTutorialEvent(tag: tag)
        }

        /*
         * 목표달성
         * @param tag: tag의 경우, 한 광고계정 내에서 같은 이벤트코드 내에서도 특별한 분류가 필요할 때 추가합니다.
         */
        Function("sendMissionCompleteEvent") { (tag: String?) in
            KakaoAdTracker.sendMissionCompleteEvent(tag: tag)
        }        
    }
}

File: android/src/main/java/expo/modules/kakaoad/ExpoKakaoAdModule.kt
package expo.modules.kakaoad

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.kakao.ad.tracker.KakaoAdTracker
import com.kakao.ad.common.json.*
import com.kakao.ad.tracker.send
import java.util.Currency
import java.util.Locale

class ExpoKakaoAdModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoKakaoAd")

    /**
     * KakaoAdTracker를 초기화합니다.
     */
    AsyncFunction("activate") { promise: Promise ->
      val ctx = appContext.reactContext?.applicationContext
      if (ctx != null && !KakaoAdTracker.isInitialized) {
        val resId = ctx.resources.getIdentifier(
          "kakao_ad_track_id", "string", ctx.packageName
        )
        if (resId != 0) {
          val trackId = ctx.getString(resId)
          KakaoAdTracker.init(ctx, trackId)
        }
      }
      promise.resolve(null)
    }

    /**
     * 가입완료 이벤트(CompleteRegistration)를 전송합니다.
     */
    Function("sendCompleteRegistrationEvent") { tag: String? ->
      val event = CompleteRegistration().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 로그인 이벤트(Login)를 전송합니다.
     */
    Function("sendLoginEvent") { tag: String? ->
      val event = Login().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 검색 이벤트(Search)를 전송합니다.
     */
    Function("sendSearchEvent") { tag: String?, keyword: String? ->
      val event = Search().apply {
        this.tag = tag ?: ""
        this.search_string = keyword ?: ""
      }
      event.send()
    }

    /**
     * 콘텐츠/상품 조회 이벤트(ViewContent)를 전송합니다.
     */
    Function("sendViewContentEvent") { tag: String?, contentId: String? ->
      val event = ViewContent().apply {
        this.tag = tag ?: ""
        this.content_id = contentId ?: ""
      }
      event.send()
    }

    /**
     * 관심상품 추가 이벤트(AddToWishList)를 전송합니다.
     */
    Function("sendAddToWishListEvent") { tag: String?, contentId: String? ->
      val event = AddToWishList().apply {
        this.tag = tag ?: ""
        this.content_id = contentId ?: ""
      }
      event.send()
    }

    /**
     * 장바구니 추가 이벤트(AddToCart)를 전송합니다.
     */
    Function("sendAddToCartEvent") { tag: String?, contentId: String? ->
      val event = AddToCart().apply {
        this.tag = tag ?: ""
        this.content_id = contentId ?: ""
      }
      event.send()
    }

    /**
     * 장바구니 보기 이벤트(ViewCart)를 전송합니다.
     */
    Function("sendViewCartEvent") { tag: String? ->
      val event = ViewCart().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 구매 이벤트(Purchase)를 전송합니다.
     */
    Function("sendPurchaseEvent") { 
      tag: String?, 
      totalQuantity: Int, 
      totalPrice: Double, 
      currency: String?,
      products: List<Map<String, Any?>>
    ->
      val event = Purchase().apply { this.tag = tag ?: "" }

      event.products = products.map { record ->
        Product().also { product ->
          product.id = (record["id"] as? String) ?: "" // 상품 ID
          product.name = (record["name"] as? String) ?: "" // 상품명
          product.quantity = (record["quantity"] as? Double)?.toInt() ?: 0 // 개수
          product.price = (record["price"] as? Double) ?: 0.0 // 금액
        }
      }
      // currency가 넘겨지지 않으면 기본으로 KRW 사용
      event.currency = if (currency != null) {
        Currency.getInstance(currency)
      } else {
        Currency.getInstance(Locale.KOREA)
      }
      event.total_quantity = totalQuantity
      event.total_price = totalPrice
      event.send()
    }

    /**
     * 인앱 구매 이벤트(InAppPurchase)를 전송합니다.
     */
    Function("sendInAppPurchaseEvent") { 
      tag: String?, 
      totalQuantity: Int, 
      totalPrice: Double, 
      currency: String?,
      products: List<Map<String, Any?>>
    ->
      val event = InAppPurchase().apply { this.tag = tag ?: "" }

      event.products = products.map { record ->
        Product().also { product ->
          product.id = (record["id"] as? String) ?: "" // 상품 ID
          product.name = (record["name"] as? String) ?: "" // 상품명
          product.quantity = (record["quantity"] as? Double)?.toInt() ?: 0 // 개수
          product.price = (record["price"] as? Double) ?: 0.0 // 금액
        }
      }
      event.currency = if (currency != null) {
        Currency.getInstance(currency)
      } else {
        Currency.getInstance(Locale.KOREA)
      }
      event.total_quantity = totalQuantity
      event.total_price = totalPrice
      event.send()
    }

    /**
     * 잠재고객 이벤트(Participation)를 전송합니다.
     *
     * 잠재고객([Participation]) 이벤트는 아래 태그([Participation.tag])를 추가 설정하면 전환을 최적화하는데 도움이 됩니다.
     *
     * 권장 태그 추가 목적 (태그값):
     * - 사전예약 (PreBooking)
     * - 상담신청 (Consulting)
     * - 시승신청 (DrivingTest)
     * - 대출한도조회 (LoanLimitCheck)
     * - 보험료조회 (InsuranceCheck)
     */
    Function("sendParticipationEvent") { tag: String? ->
      val event = Participation().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 사전준비 이벤트(Preparation)를 전송합니다.
     */
    Function("sendPreparationEvent") { tag: String? ->
      val event = Preparation().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 가입 및 등록 이벤트(SignUp)를 전송합니다.
     *
     * 가입 및 등록([SignUp]) 이벤트는 아래 태그([SignUp.tag])를 추가 설정하면 전환을 최적화하는데 도움이 됩니다.
     *
     * 권장 태그 추가 목적 (태그값):
     * - 서비스가입 (SignUp)
     * - 구독완료 (Subscription)
     * - 카드발급 (CardIssuance)
     * - 계좌개설 (OpeningAccount)
     * - 대출신청 (LoanApplication)
     */
    Function("sendSignUpEvent") { tag: String? ->
      val event = SignUp().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 튜토리얼 이벤트(Tutorial)를 전송합니다.
     */
    Function("sendTutorialEvent") { tag: String? ->
      val event = Tutorial().apply { this.tag = tag ?: "" }
      event.send()
    }

    /**
     * 미션완료 이벤트(MissionComplete)를 전송합니다.
     */
    Function("sendMissionCompleteEvent") { tag: String? ->
      val event = MissionComplete().apply { this.tag = tag ?: "" }
      event.send()
    }
  }
}